File: Zeze/Raft/RocksRaft/LogMap2.cs
﻿using System;
using System.Collections.Generic;
using System.Text;
using Zeze.Serialize;

namespace Zeze.Raft.RocksRaft
{
	public class LogMap2<K, V> : LogMap1<K, V>
		where V : Bean, new()
	{
		// changed V logs. using in collect.
		public ISet<LogBean> Changed { get; } = new HashSet<LogBean>();

		// changed with key. using in encode/decode FollowerApply
		public Dictionary<K, LogBean> ChangedWithKey { get; } = new Dictionary<K, LogBean>();

		internal override Log BeginSavepoint()
		{
			var dup = new LogMap2<K, V>();
			dup.Belong = Belong;
			dup.VariableId = VariableId;
			dup.Value = Value;
			return dup;
		}

		public override void Decode(ByteBuffer bb)
        {
			ChangedWithKey.Clear();
			for (int i = bb.ReadUInt(); i > 0; --i)
			{
				var key = SerializeHelper<K>.Decode(bb);
				var value = SerializeHelper<LogBean>.Decode(bb);
				ChangedWithKey.Add(key, value);
			}
            base.Decode(bb);
        }

        public override void Encode(ByteBuffer bb)
        {
			if (null != Value)
            {
				foreach (var c in Changed)
				{
					if (CollMap2<K, V>.PropertyMapKey != null)
					{
						var pkey = (K)CollMap2<K, V>.PropertyMapKey.GetValue(c.This);
						if (false == Putted.ContainsKey(pkey) && false == Removed.Contains(pkey))
							ChangedWithKey.Add(pkey, c);
						continue;
					}
					// slow search.
					foreach (var e in Value)
					{
						if (c.Belong == e.Value)
						{
							if (false == Putted.ContainsKey(e.Key) && false == Removed.Contains(e.Key))
								ChangedWithKey.Add(e.Key, c);
							break;
						}
					}
				}
			}
			
			bb.WriteUInt(ChangedWithKey.Count);
			foreach (var e in ChangedWithKey)
            {
				SerializeHelper<K>.Encode(bb, e.Key);
				SerializeHelper<LogBean>.Encode(bb, e.Value);
			}
			base.Encode(bb);
        }

		public override void Collect(Changes changes, Bean recent, Log vlog)
		{
			if (Changed.Add((LogBean)vlog))
            {
				changes.Collect(recent, this);
			}
		}

		public override string ToString()
		{
			var sb = new StringBuilder();
			sb.Append(" Putted:");
			ByteBuffer.BuildString(sb, Putted);
			sb.Append(" Removed:");
			ByteBuffer.BuildString(sb, Removed);
			sb.Append(" Changed:");
			ByteBuffer.BuildString(sb, Changed);
			return sb.ToString();
		}
	}
}

File: ZezeJava/ZezeJava/src/main/java/Zeze/Raft/RocksRaft/LogMap2.java
package Zeze.Raft.RocksRaft;

import java.lang.invoke.MethodHandle;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Set;
import Zeze.Serialize.ByteBuffer;
import Zeze.Serialize.SerializeHelper;
import Zeze.Util.Reflect;

public class LogMap2<K, V extends Bean> extends LogMap1<K, V> {
	private final Set<LogBean> Changed = new HashSet<>(); // changed V logs. using in collect.
	private final HashMap<K, LogBean> ChangedWithKey = new HashMap<>(); // changed with key. using in encode/decode FollowerApply
	private final MethodHandle valueFactory;

	public LogMap2(Class<K> keyClass, Class<V> valueClass) {
		super("Zeze.Raft.RocksRaft.LogMap2<" + Reflect.GetStableName(keyClass) + ", "
				+ Reflect.GetStableName(valueClass) + '>', keyClass, valueClass);
		valueFactory = Reflect.getDefaultConstructor(valueClass);
	}

	LogMap2(int typeId, SerializeHelper.CodecFuncs<K> keyCodecFuncs, MethodHandle valueFactory) {
		super(typeId, keyCodecFuncs, null);
		this.valueFactory = valueFactory;
	}

	public final Set<LogBean> getChanged() {
		return Changed;
	}

	public final HashMap<K, LogBean> getChangedWithKey() {
		return ChangedWithKey;
	}

	@Override
	public Log BeginSavepoint() {
		var dup = new LogMap2<K, V>(getTypeId(), keyCodecFuncs, valueFactory);
		dup.setBelong(getBelong());
		dup.setVariableId(getVariableId());
		dup.setValue(getValue());
		return dup;
	}

	@SuppressWarnings("unchecked")
	@Override
	public void Encode(ByteBuffer bb) {
		if (getValue() != null) {
			for (var c : Changed) {
				Object pkey = c.getThis().getMapKey();
				//noinspection SuspiciousMethodCalls
				if (!getPutted().containsKey(pkey) && !getRemoved().contains(pkey))
					ChangedWithKey.put((K)pkey, c);
			}
		}
		bb.WriteUInt(ChangedWithKey.size());
		var keyEncoder = keyCodecFuncs.encoder;
		for (var e : ChangedWithKey.entrySet()) {
			keyEncoder.accept(bb, e.getKey());
			e.getValue().Encode(bb);
		}

		// super.Encode(bb);
		bb.WriteUInt(getPutted().size());
		for (var p : getPutted().entrySet()) {
			keyEncoder.accept(bb, p.getKey());
			p.getValue().Encode(bb);
		}
		bb.WriteUInt(getRemoved().size());
		for (var r : getRemoved())
			keyEncoder.accept(bb, r);
	}

	@SuppressWarnings("unchecked")
	@Override
	public void Decode(ByteBuffer bb) {
		ChangedWithKey.clear();
		var keyDecoder = keyCodecFuncs.decoder;
		for (int i = bb.ReadUInt(); i > 0; i--) {
			var key = keyDecoder.apply(bb);
			var value = new LogBean();
			value.Decode(bb);
			ChangedWithKey.put(key, value);
		}

		// super.Decode(bb);
		getPutted().clear();
		for (int i = bb.ReadUInt(); i > 0; i--) {
			var key = keyDecoder.apply(bb);
			V value;
			try {
				value = (V)valueFactory.invoke();
			} catch (Throwable e) {
				throw new RuntimeException(e);
			}
			value.Decode(bb);
			getPutted().put(key, value);
		}
		getRemoved().clear();
		for (int i = bb.ReadUInt(); i > 0; i--)
			getRemoved().add(keyDecoder.apply(bb));
	}

	@Override
	public void Collect(Changes changes, Bean recent, Log vlog) {
		if (Changed.add((LogBean)vlog))
			changes.Collect(recent, this);
	}

	@Override
	public String toString() {
		var sb = new StringBuilder();
		sb.append(" Putted:");
		ByteBuffer.BuildSortedString(sb, getPutted());
		sb.append(" Removed:");
		ByteBuffer.BuildSortedString(sb, getRemoved());
		sb.append(" Changed:");
		ByteBuffer.BuildSortedString(sb, Changed);
		return sb.toString();
	}
}
